### ISSUE TITLE
OpenMoji heatmap "status" viz

### ISSUE THREAD

**Issue by axelpale** [hfg-gmuend/openmoji#30]
I think OpenMoji project is awesome. However, I find it a bit difficult to get a grasp of the amount and variety of the great work the project has accumulated this far. The bank of over 3000 emojis is rather tedious to visualise and browse through to get this grasp.

Therefore, I wanted to give something back and support the project on that part. I created a portable way to represent the full variety of OpenMoji emojis in a fun way. It is an animation where a continuous stream of randomly picked emojis fall down like the flakes of snow. You can check it out at:

http://demo.akselipalen.com/mojifall/

In addition to being embeddable into a web page, this could serve as a background animation wherever you promote the OpenMoji project, seminars for example.

For a bit of technical side, I had to overcome a challenge of how to make the animation efficient for a quick page load. There was about 3180 emojis in the OpenMoji package I used in the animation. These amount to about 13 MB which is too slow to load all at once. To make things tricky, HTML Canvas, the way the animation is drawn, requires the rendered images to be downloaded beforehand. The solution was not trivial to find and put in code but now it works well and seems to be quite bandwidth-efficient. To start, the animation requires a browser to download about 60 kB. To keep the animation running and emojis falling, additional 25 kB needs to be downloaded per second. For reference, the front page of OpenMoji.org sums up to 3.4 MB.

Let me know if you find it useful in some way. I am happy to adjust it for the needs of the project and share the source.

**Comment by dnlutz** [comment]
lovely … you can be sure that we will use it ; ) we already thought about something like an "openmoji heatmap", but i really like your more playful approach. thank you very much for your contribution!

**Comment by axelpale** [comment]
Great to hear!

Your heatmap approach gave me an idea: I am interested to create another, bit different animation. The new one would reflect the real-world usage of emojis by having a similar frequency distribution. If you are aware of a representative data set of global or local use of emojis, I am all ears.

**Comment by dnlutz** [comment]
http://emojitracker.com/ might be a suitable source?

**Comment by axelpale** [comment]
Oh, awesome! I will look into that. Umm... their front page lists only about 850 emojis what is much less than OpenMoji's 3180. I assume the difference is due to the absence of color variants in emojitracker's feed. Please feel free to correct me if there is something else causing the big difference in the numbers.

**Comment by axelpale** [comment]
For later reference, here is a direct url to Emojitrackers emoji score distribution API:
https://api.emojitracker.com/v1/rankings

I will use that as the source frequency distribution.

**Comment by axelpale** [comment]
Here it is: the animation that mimics real-world emoji usage in Twitter! Also, I published the source codes, please see below.

Animation:
- [demo.akselipalen.com/mojifall-openmoji-emojitracker](http://demo.akselipalen.com/mojifall-openmoji-emojitracker/)

It is rather touching to think that for each falling emoji there was a real emotion behind it... felt by somebody... somewhere.

Repositories:
- See [github.com/axelpale/mojifall](https://github.com/axelpale/mojifall) for the original unweighted animation.
- See [github.com/axelpale/mojifall/tree/emojitracker](https://github.com/axelpale/mojifall/tree/emojitracker) for the new realistically weighted animation.

**Comment by dnlutz** [comment]
Very nice contribution, it really looks like visual poetry!

**Comment by dnlutz** [comment]
For the future we might also need an unexciting heatmap like this to tackle status, changes or tasks. Maybe the Emoji could be displayed very small or with a tool tip like enlargment,
![emoji_heatmap](https://user-images.githubusercontent.com/38292227/66044770-97cdb380-e522-11e9-8d15-44e3e92a2a0c.png)

**Comment by axelpale** [comment]
I see. Such a heatmap is very well doable. The main challenge, as I see it, is the retrieval of the data.

Possible data sources and some pros and cons for each:
- /openmoji/data/openmoji.csv
  - processing complexity: easy; heatmap app just needs to read the file
  - no status column available currently
  - status would be tedius to keep up to date manually
- GitHub issues
  - processing complexity: high; issues would need some keywords for the data processor to recognize
  - already very visible at [openmoji/issues](https://github.com/hfg-gmuend/openmoji/issues)
  - only a few issues open currently; low benefit versus the amount of work
  - main benefit: the correct source for tasks
- Git change log
  - processing complexity: medium
  - changed files that are emojis are easy to recognize
  - provides only change history – no status or task data

What are your thoughts about the data retrieval?

**Comment by b-g** [comment]
Hi @dnlutz, can you elaborate on the intention of the "heatmap"? Would it no be mainly of use if an new unicode emoji version comes in and we want to show the delta between what we already have vs. what is new / what has to be done.

If so, it would be easy to get the data basically `npm emojibase-data of e.g. Emoji 13` minus `/openmoji/data/openmoji.json` = the diff is what don't have or what we did in addition.

**Comment by dnlutz** [comment]
I think a heatmap could be useful for many purposes, for example the "snow man issue":
https://github.com/hfg-gmuend/openmoji/issues/31
- We could mark all emoji that have a problem with dark backgrounds
- Emoji which are really tough to fix could be marked with a red spot

Another usecase could be the overall improvement of existing emoji. OpenMoji is an open source project mainly created by students. I am proud of what we achieved but we still have mixed quaility in our emoji: most of them are great, some are o.k. and some still need refinement. A heatmap (together with little notes) could be a powerful tool to communicate where we see room for improvement. And if we could link the notes to different sections of our styleguide things would become very transparent: which emoji should be revised and why?

The next question is the following: How can we reach willing designers for these tasks? I am really happy and grateful that people start to work with us on OpenMoji who are NOT from our university (like you @axelpale and @felixhaeberle ; ) But I have the vague feeling that on GitHub we mainly attract coders … but we also need more designers to improve and evolve our Emoji system.

**Comment by felixhaeberle** [comment]
I'm not introduced in this topic here but because of my mention: I'll join the university next week. :)

**Comment by dnlutz** [comment]
Interesting ; ) Schwäbisch Gmünd? IG? IoT?

**Comment by felixhaeberle** [comment]
IG ✌️

**Comment by b-g** [comment]
@felixhaeberle :) Welcome at HfG!!

**Comment by axelpale** [comment]
@dnlutz @b-g I will try to come up with a heatmap prototype based on b-g's proposition to use `emojibase-data`. Keeping it simple is a good way forward.

**Comment by b-g** [comment]
@axelpale @dnlutz, yes such a heatmap or maybe better a "status" viz would be helpful. But from a technical point of view I don't like the idea to add that status to the `data/openmoji.*` files. 

Another way could be if @dnlutz could simply track the status in a gSheet and we pull the data from gDrive?

**Comment by dnlutz** [comment]
i'm not sure wether it makes sense to start the heat map right now. we don't have a clear roadmap for improvements and refinements yet. the next update will hopefully take place in the private use are (beyond unicode ; )

**Comment by Sam0711er** [comment]
Super late to the party, but welcome to @hfg-gmuend dear @felixhaeberle ! ✌️

**Comment by axelpale** [comment]
Hi everyone.

I try to be a man of my word so here is a prototype that compares Emojibase data to OpenMoji data. The comparison ended up to be rather uninformative as OpenMoji seems to already implement all of the emojis listed in Emojibase (Emoji 12, Unicode 12). Nevertheless, you can check it out at:

http://demo.akselipalen.com/openmoji-status/

Here are some notes I gathered during implementation:

- Emojibase provides an alternative grouping by [Emoji and Unicode versions](https://milesj.gitbook.io/emojibase/data). This grouping could be informative when OpenMoji needs to update Emoji/Unicode version 13.
- I believe the ability for contributor to reference to exact piece of work helps them to communicate. OpenMoji provides direct links to certain emojis, like [this](https://openmoji.org/library/#emoji=1F600). However, I did not find a way to construct direct links to skin-tone variants. Therefore all toned emojis in the prototype link to their base emoji instead.
- Emojibase releases its data via a npm package. OpenMoji does not. This makes usage a bit harder for developers because they need to fetch the data from OpenMoji's repository instead of installing the data as a project dependency.
- The prototype uses the default emoji set of the browser as icons. The richness of this set seems to change from device to device, causing unsupported emojis show more or less as boxes.

**Comment by dnlutz** [comment]
thank you very much @axelpale … that's exactly how I imagined it!

we definitely need such a heat map / compact overview for future developments.
as i'm a coding layman i can't really evaluate what you wrote about npm etc. 
(but it sounds quite reasonable ; )

in the next update we will add some stuff in the private use area but hopefully we can use your contribution for the unicode 13 update next year!

**Comment by b-g** [comment]
Thank you @axelpale! Also a few comments:

> I believe the ability for contributor to reference to exact piece of work helps them to communicate. OpenMoji provides direct links to certain emojis, like this. However, I did not find a way to construct direct links to skin-tone variants. Therefore all toned emojis in the prototype link to their base emoji instead.

We are aware of direct links and should shortly be possible for skin-tone variants

> Emojibase releases its data via a npm package. OpenMoji does not.

Yes would be nice to ship OpenMoji with npm ... but as every release involves quite some manual work we never could agree what npm install openmoji would actually distribute. Just the src files, or just the svg files in color and black?

> The prototype uses the default emoji set of the browser as icons. The richness of this set seems to change from device to device, causing unsupported emojis show more or less as boxes.

Why not simply use the OpenMoji 72x72 png files via a direct link e.g. https://hfg-gmuend.github.io/openmoji/color/72x72/1F600.png? I think I would favor this over the emoji system font of a user.

**Comment by axelpale** [comment]
Thanks for comments! To answer @b-g's question:

> Why not simply use the OpenMoji 72x72 png files via a direct link e.g. https://hfg-gmuend.github.io/openmoji/color/72x72/1F600.png? I think I would favor this over the emoji system font of a user.

(I aimed to give a short answer but the topic brought up a general deliverability challenge for OpenMoji which might require its own issue thread. Therefore I will discuss it with additional detail.)

The reason in using browser's native emojis is a technical kind. There is a practical upper limit how many [HTTP requests](https://www.w3schools.com/tags/ref_httpmethods.asp) a webpage should do to be complete. A single page making about 1000 requests, one for each emoji image file, is **not feasible** except during local development phase. Every HTTP request adds a big overhead  to the amount of data transmitted and thus web browsers limit the number of concurrent requests to three or five and queue the rest. In addition, a large number of requests places a large load on the server infrastructure. Therefore web developers try to keep the number of requests as low as possible.

I faced this limitation during the early iterations of [Mojifall demo](http://demo.akselipalen.com/mojifall/) where the browser had to download the full set of over 3000 emojis at the initial page load. This forced me to improve [the underlying animation package](https://www.npmjs.com/package/sprinkler) so that it downloads emojis in a lazy manner, only after an emoji is truly needed. The [library](https://openmoji.org/library/) portion of OpenMoji web site approaches the problem in a similar way and downloads additional emojis on the fly as the user scrolls down the list.

Web developers approach this issue in three ways. The first is to load images only when needed, as mentioned above. The second and the third way recognise that it is much easier to move a box of Legos than every piece individually.

The second way uses **image sprites** where multiple images are composed into one single large image. To render an individual piece, the devs use CSS styles to crop the needed portion of the big image. As the result, only one image file needs to be downloaded, meaning only one HTTP request needs to be made by the browser. Although the image file might be large, it will download quickly in one stream. See [here](https://css-tricks.com/css-sprites/) and [here](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Images/Implementing_image_sprites_in_CSS) for further introduction.

**True type fonts** provide the third approach. OpenMoji seems to already have one ttf available for download. However, [the font file](https://github.com/hfg-gmuend/openmoji/blob/master/font/OpenMoji-Color.ttf) is 11.7 megabytes in size which is too much for a web page although better than thousands of HTTP requests.

An optimal solution from a web developer's perspective would be a set of pre-made low resolution sprite images or a true type font with a footprint much smaller than the current one. These containers could be split group-wise to allow developers of different apps to find balance between file sizes and number of files to download.

This deliverability challenge might deserve its own issue thread. I believe that the use case where the client needs to download the full emoji set is relatively common. Comment forms, article editors, and all other apps where an emoji palettes begin to be ubiquitous can integrate OpenMoji only if the emoji set is practically downloadable.
